### PR TITLE
Add parsing error message field to TrinoQueryProperties

### PR DESCRIPTION
**Pull request by Chaho12**
## Description

Add a field (private Optional<String> ParseErrorMessage) to TrinoQueryProperties class that stores `e.getMessage()` when there is error.
https://github.com/trinodb/trino-gateway/blob/main/gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java#L213-L224

We have server connected with gateway that runs explain before actually running the query.
But not all queries need to be sent with explain.

For example there are cases where queries with parsing failure (no need to run explain cause it is highly likely to fail anyway due to syntax error)
- we need to tell user on WHY query has failed, but running another sql parser or explain is waste of resource
- since gateway already does query parsing, why not use that failure info.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

### CODE REVIEW

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 1]
```diff
@@ -519,6 +528,12 @@ public boolean isQueryParsingSuccessful()
         return isQueryParsingSuccessful;
     }
 
+    @JsonProperty
+    public Optional<String> getParseErrorMessage()
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:532]: This method is used only from tests. What's the motivation of this change?

**Review by ebyhr — COMMENTED** [review]
> (x) Release notes are required, with the following suggested text:
> * Add parsing error message field to TrinoQueryProperties

Why do we need to mention this change in a release note?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 1 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:532]: We have server connected with gateway that runs explain before actually running the query. 
But not all queries need to be sent with explain.

For example there are cases where queries with parsing failure (no need to run explain cause it is highly likely to fail anyway due to syntax error)
  - we need to tell user on WHY query has failed, but running another sql parser or explain is waste of resource
  - since gateway already does query parsing, why not use that failure info.

**Comment by Chaho12** [comment]
> Why do we need to mention this change in a release note?

Thought there might be some people like me who would need the error. 
Not necessary though, I could just update the docs instead.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 2]
```diff
@@ -114,7 +115,8 @@ public TrinoQueryProperties(
             @JsonProperty("schemas") Set<String> schemas,
             @JsonProperty("catalogSchemas") Set<String> catalogSchemas,
             @JsonProperty("isNewQuerySubmission") boolean isNewQuerySubmission,
-            @JsonProperty("isQueryParsingSuccessful") boolean isQueryParsingSuccessful)
+            @JsonProperty("isQueryParsingSuccessful") boolean isQueryParsingSuccessful,
+            @JsonProperty("parseErrorMessage") Optional<String> parseErrorMessage)
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:119]: Why do we need to include "parse" in the field name? Should we add another variable if we find different error cause?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 3]
```diff
@@ -212,14 +215,17 @@ else if (statement instanceof ExecuteImmediate executeImmediate) {
         }
         catch (IOException e) {
             log.warn("Error extracting request body for rules processing: %s", e.getMessage());
+            parseErrorMessage = Optional.ofNullable(e.getMessage());
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:218]: This isn't a parse error because IOException happens before parsing. The variable should be renamed or we shouldn't set the variable here.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 4]
```diff
@@ -379,8 +386,10 @@ private void setCatalogAndSchemaNameFromSchemaQualifiedName(
 
     private RequestParsingException unsetDefaultExceptionSupplier()
     {
+        String errorString = "Name not fully qualified";
         isQueryParsingSuccessful = false;
-        return new RequestParsingException("Name not fully qualified");
+        parseErrorMessage = Optional.of(errorString);
+        return new RequestParsingException(errorString);
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:392]: `parseErrorMessage` is set twice here and catch block of `processRequestBody`.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 5]
```diff
@@ -212,14 +215,17 @@ else if (statement instanceof ExecuteImmediate executeImmediate) {
         }
         catch (IOException e) {
             log.warn("Error extracting request body for rules processing: %s", e.getMessage());
+            parseErrorMessage = Optional.ofNullable(e.getMessage());
             isQueryParsingSuccessful = false;
         }
         catch (ParsingException e) {
             log.info("Could not parse request body as SQL: %s; Message: %s", body, e.getMessage());
+            parseErrorMessage = Optional.ofNullable(e.getMessage());
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:223]: Is `e.getMessage()` really nullable?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 6]
```diff
@@ -212,14 +215,17 @@ else if (statement instanceof ExecuteImmediate executeImmediate) {
...
             isQueryParsingSuccessful = false;
         }
         catch (RequestParsingException e) {
             log.warn(e, "Error parsing request for rules");
+            parseErrorMessage = Optional.ofNullable(e.getMessage());
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:228]: Is `e.getMessage()` really nullable?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 7]
```diff
@@ -97,6 +97,7 @@ public class TrinoQueryProperties
     private Set<String> catalogSchemas = ImmutableSet.of();
     private boolean isNewQuerySubmission;
     private boolean isQueryParsingSuccessful;
+    private Optional<String> parseErrorMessage;
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:100]: Why do we need both isQueryParsingSuccessful and parseErrorMessage variables? Is it possible to remove isQueryParsingSuccessful and return the value by checking the size of parseErrorMessage on the fly?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 1 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:532]: > since gateway already does query parsing, why not use that failure info.

I don't think relying on Trino Gateway's parser is safe. For instance, the behavior might be different from the backend Trino because of version mismatch.

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:532]: Sure updated description.

> I don't think relying on Trino Gateway's parser is safe. For instance, the behavior might be different from the backend Trino because of version mismatch.

That is true. Gateway does not guarantee that, but I think we should leave this up to the users to decide on how much they rely on the result. 
Btw in our case, we just use `isQueryParsingSuccessful` to see if we will run explain to fetch IO Plan or not, as it is part of our routing rules. (There are several steps on deciding which routing group to route, but one heavy step of running explain can be skipped). 
This failure info is a supplementary info where we tell users that query parsing has failed so please do check your sql 
- As of now, we use sql parser libraries to see why parsing has failed, which is also not that accurate.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 7 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:100]: So if there isn't an errorMessage than it is considered successful? 
Hmm.. I prefer having a separate variable. @willmostly any thoughts?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 2 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:119]: Making it a more general field as `errorMessage` and storing any sort of errors here sounds good.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 5 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:223]: Unless we have a logical bug, not really. 
Since parseErrorMessage is type `Optional<String>`, i couldn't directly save it as e.getMessage().
Any better ideas?

Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:223]: Why not use `Optional.of()` method?

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 7 above)
Comment (ebyhr) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:100]: When does isQueryParsingSuccessful become true when parseErrorMessage is present?
When does isQueryParsingSuccessful become false when parseErrorMessage is empty? 

Having two variables is troublesome as it's easy to miss updating another one. Why do you prefer having a separate variable?

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:100]: Nowadays am used to using Golang as it always has error field to check before doing any further actions. 
- https://go.dev/blog/error-handling-and-go

Quite convenient should i say. 
But am ok with just having error message and removing boolean field.

**Comment by mosabua** [comment]
Looks like @ebyhr and @Chaho12 are closing this out together shortly. Let me know if you want a review or other help from me.

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java (hunk 7 above)
Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java]: Any ideas on handling this? because when i do this
```java
    @JsonProperty("isQueryParsingSuccessful")
    public boolean isQueryParsingSuccessful()
    {
        return errorMessage.isEmpty();
    }
```
I get this error. As of now, isQueryParsingSuccessful is assigned but never accessed. 
There is no need to access the field in code, but i don't see any other better ways than simply adding @SuppressWarnings("unused")
```
gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:99:21
java: [UnusedVariable] The field 'isQueryParsingSuccessful' is never read.
```

File: gateway-ha/src/test/java/io/trino/gateway/ha/router/TestTrinoQueryProperties.java [hunk 8]
```diff
@@ -53,6 +54,7 @@ void testJsonCreator()
         assertThat(deserializedTrinoQueryProperties.getCatalogs()).isEqualTo(trinoQueryProperties.getCatalogs());
         assertThat(deserializedTrinoQueryProperties.getCatalogSchemas()).isEqualTo(trinoQueryProperties.getCatalogSchemas());
         assertThat(deserializedTrinoQueryProperties.isNewQuerySubmission()).isEqualTo(trinoQueryProperties.isNewQuerySubmission());
-        assertThat(deserializedTrinoQueryProperties.isQueryParsingSuccessful()).isEqualTo(trinoQueryProperties.isQueryParsingSuccessful());
+        assertThat(deserializedTrinoQueryProperties.isQueryParsingSuccessful()).isEqualTo(trinoQueryProperties.isQueryParsingSuccessful());;
```
Comment (ebyhr) [inline gateway-ha/src/test/java/io/trino/gateway/ha/router/TestTrinoQueryProperties.java:57]:
```suggestion
        assertThat(deserializedTrinoQueryProperties.isQueryParsingSuccessful()).isEqualTo(trinoQueryProperties.isQueryParsingSuccessful());
```

File: gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java [hunk 9]
```diff
@@ -114,7 +116,8 @@ public TrinoQueryProperties(
             @JsonProperty("schemas") Set<String> schemas,
             @JsonProperty("catalogSchemas") Set<String> catalogSchemas,
             @JsonProperty("isNewQuerySubmission") boolean isNewQuerySubmission,
-            @JsonProperty("isQueryParsingSuccessful") boolean isQueryParsingSuccessful)
+            @JsonProperty("isQueryParsingSuccessful") boolean isQueryParsingSuccessful,
```
Comment (ebyhr) [inline gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java:119]: Why do we still need this field? I thought just having the getter is sufficient.

Comment (Chaho12) [reply gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoQueryProperties.java]: @ebyhr PTAL if I understood correctly, when you mean just having getter.
